cura/Settings/MachineManager.py

```diff
@@ -1496,3 +783,268 @@ def setSettingForAllExtruders(self, setting_name: str, property_name: str, prope
+        return self._current_root_material_id
+
+    @pyqtProperty("QVariant", notify = rootMaterialChanged)
+    def currentRootMaterialName(self):
+        # initial filling the current_root_material_name
+        if self._global_container_stack:
+            self._current_root_material_name = {}
+            for position in self._global_container_stack.extruders:
+                if position not in self._current_root_material_name:
+                    material = self._global_container_stack.extruders[position].material
+                    self._current_root_material_name[position] = material.getName()
+        return self._current_root_material_name
+
+    ##  Return the variant names in the extruder stack(s).
+    ##  For the variant in the global stack, use activeVariantBuildplateName
+    @pyqtProperty("QVariant", notify = activeVariantChanged)
+    def activeVariantNames(self):
+        result = {}
+
+        active_stacks = ExtruderManager.getInstance().getActiveExtruderStacks()
+        if active_stacks is not None:
+            for stack in active_stacks:
+                variant_container = stack.variant
+                position = stack.getMetaDataEntry("position")
+                if variant_container and variant_container != self._empty_variant_container:
+                    result[position] = variant_container.getName()
+
+        return result
+
+    def _setEmptyQuality(self):
+        self._current_quality_group = None
+        self._current_quality_changes_group = None
+        self._global_container_stack.quality = self._empty_quality_container
+        self._global_container_stack.qualityChanges = self._empty_quality_changes_container
+        for extruder in self._global_container_stack.extruders.values():
+            extruder.quality = self._empty_quality_container
+            extruder.qualityChanges = self._empty_quality_changes_container
+
+        self.activeQualityGroupChanged.emit()
+        self.activeQualityChangesGroupChanged.emit()
+
+    def _setQualityGroup(self, quality_group, empty_quality_changes = True):
+        self._current_quality_group = quality_group
+        if empty_quality_changes:
+            self._current_quality_changes_group = None
+
+        # Set quality and quality_changes for the GlobalStack
+        self._global_container_stack.quality = quality_group.node_for_global.getContainer()
+        if empty_quality_changes:
+            self._global_container_stack.qualityChanges = self._empty_quality_changes_container
+
+        # Set quality and quality_changes for each ExtruderStack
+        for position, node in quality_group.nodes_for_extruders.items():
+            self._global_container_stack.extruders[position].quality = node.getContainer()
+            if empty_quality_changes:
+                self._global_container_stack.extruders[position].qualityChanges = self._empty_quality_changes_container
```

Comment:
I'm seeing this pattern of unpacking a group to the appropriate stacks more often. We could consider creating a function on the groups that you can provide a global stack and extruder stacks and it'll put the containers in their place, or even to create a function on the stacks that accept a group and picks the correct profile from them.

Reply:
Leave it like this for now.